contracts/src/Teleporter/ReceiptQueue.sol

```diff
@@ -66,32 +69,31 @@ contract ReceiptQueue {
      * - `msg.sender` must be the owner.
      * - The queue must be non-empty.
      */
-    function dequeue()
-        external
+    function dequeue(
+        TeleporterMessageReceiptQueue storage queue
+    )
+        internal
         returns (TeleporterMessageReceipt memory result)
     {
-        if (msg.sender != owner) {
-            revert Unauthorized();
+        unchecked {
```

Comment:
What are the gas savings of using `unchecked` here vs the default checked arithmetic? Have you tried using `SafeMath`? Overflow is still theoretically possible here, even though it's practically impossible.

Reply:
As of Solidity 0.8.0, the compiler itself takes care of overflow and underflow checks, so `SafeMath` shouldn't be needed here.

It's still worth thinking through if we should allow it or not though (even though it would only occur once more than 2^256 messages have been received by a subnet 😅). If we don't allow overflow, I believe there is a hard limit of 2^256 messages that could ever be received by a single `TeleporterMessenger` instance, because all subsequent messages would revert here when they are tried to be delivered. If we do allow overflow, then after the 2^256th message is received, the next message will loop around and be stored at slot 0. This could be okay, if by that time slot 0 was freed up by the receipt being removed from the queue. If it had yet to be removed, we would overwrite it, and the `first` and `last` values would become equal, effectively making the queue empty. 

While it's unfeasible for that to occur, there are already other places in the contract where we have a theoretical limit of 2^256 messages that could ever be sent between two subnets (because message IDs are tracked as `uint256`), so I don't think it's worth using `unchecked` here. If something overflows, lets just revert rather than possibly having it silently continue on.

(2^256 is so large that this whole thought experiment is a bit ridiculous either way. 😅)

Reply:
Agree that we should revert on overflow. If we send 1000 messages per second, it would take: 3.6712774e+64 centuries to overflow 2^256.